### DISCUSSION TITLE
Fine-tuning of Qwen-TTS for Dubbing Task

### DISCUSSION THREAD

**Discussion by sakshamgautam2001** [QwenLM/Qwen3-TTS#288] [Q&A]
We want to fine-tune the Qwen-TTS for cross-lingual voice cloning task, that is dubbing.
Where in reference we want to give audio and text in English language, and the input text will be in any european language (basically a translated version of reference text). 

the current QwenTTS problem is, when English text is given in reference and language is set for "French" in generate_voice_cloning, it does not generate voice in a proper french accent, and it looks like it is doing in-context learning directly from english. So we decided to fine-tuned it using rich dataset of dubbing from english to 4 european languages.

as done in sft_12hz.py, fine-tuning all the layers together led to noise. But fine-tuning only backbone and text_embedding layer led to something. But still the accuracy of voice didn't improved, because seems like in inference time, "text_projection" layer is being used on the text_embedding outputs, but at training time, the forward pass and gradient is calculated without the text_projection layer.
I am highly doubting on this mismatch which is leading to the same performance of the fine-tuned model as per the base model, because the features on which the model learnt were not getting appeared in inference time, due to presence of text_projection layer.

We also tried to incorporate the text_projection layer during training forward pass and gradient as well, but this led to noise again.

We have used low learning rates only (1e-6 to 5e-6), but still not able to fine-tune Qwen-TTS 12hz model. right now I am lost whether to put the text_projection layer or not during training, because at inference time text_projection layer is there, so we need to match the forward pass in the training as well, so that there will be consistency in training and inference. And removing text_projection and training the model is not leading to expected performances of model, that means the model is not using its training during inference due to the mismatch of text_projection layer.

How to handle this training, so that the model will properly learn the respective european language accent, even if the reference english text is present.

**Comment by baudcypher** [comment]
Hey — we hit almost exactly the same wall (loss goes down, output is noise / generation runs away forever), and after a lot of 
 dead ends we got a clean working recipe on the 0.6B Base. Sharing what unblocked us in case it saves you time. Your 
 text_projection intuition is partly right, but for us it wasn't the main thing — there were 4 separate bugs stacked on top of 
 each other, and any one of them alone produces "decreasing loss, garbage output."

 1. Learning rate is the #1 cause of noise / infinite generation

 The default 2e-5 is too high. We proved this with an overfit-on-1 test: train on a single clip for ~200 steps.
 - At lr=1e-4: teacher-forced loss hit 0.0000 (the model perfectly memorized the target tokens), but autoregressive generation 
 produced 40s of garbage for a 2s clip — even with do_sample=False (greedy). No EOS, ever.
 - At lr=2e-6: same loss floor, but greedy generation reproduced the clip cleanly (~2.0s, WER 0%, speaker-cosine 0.999).

 So a perfect teacher-forced loss tells you nothing about whether generation will work. High LR catastrophically forgets the 
 EOS/generation dynamics while the loss looks great. Use 2e-6.

 2. There are TWO label-shift bugs, not one

 PR #178 fixes the talker (codebook-0) double-shift, but the sub-talker loss double-shifts too and we couldn't find this 
 documented anywhere:
 - forward_sub_talker_finetune passes labels=codec_ids[:, 1:] (already shifted) into code_predictor.forward_finetune, which then 
 applies HF's ForCausalLMLoss internal shift again → codebooks 1–15 are trained on a misaligned target. They look fine in 
 training, then poison the talker's next-step input during generation.
 - Fix: ignore the returned sub_talker_loss, take sub_talker_logits and compute manual F.cross_entropy per codebook:
 sub_logits, _ = model.talker.forward_sub_talker_finetune(talker_codec_ids, talker_hidden_states)
 # sub_logits[k] predicts codebook k+1; target = talker_codec_ids[:, k+1]

 3. Vocab dimension gotcha (causes cryptic "batch_size mismatch")

 The talker's codebook-0 head outputs 3072 (codec 2048 + special tokens), but the sub-talker heads output 2048. Don't hardcode V —
 use each head's own .size(-1), or you'll get Expected input batch_size X to match target batch_size Y errors that look like a 
 data problem but aren't.

 4. speaker_encoder signature mismatch

 The installed Qwen3TTSSpeakerEncoder.forward(self, hidden_states) takes no lengths kwarg, but sft_12hz.py calls 
 model.speaker_encoder(ref_mels, lengths=ref_mel_lengths). This hard-crashes the --speaker path — which is why people end up on 
 --no_speaker (it skips this call). Fix: drop lengths= and use batch_size=1 (no padding).

 The diagnostic we wish we'd run first

 Don't trust loss. Overfit on one clip (~200 steps, lr=2e-6, both shift bugs fixed) and check that greedy generation reproduces 
 that clip — right duration, WER ~0%. If it does, your pipeline is correct and any remaining issues are data/inference. If greedy 
 runaways even with loss→0, you've still got a shift bug. This single test would have saved us days.

 Inference

 The community fork bakes the averaged speaker embedding straight into codec_embedding.weight[spk_id] inside model.safetensors, so
 at inference: load base → load_state_dict(ckpt, strict=False) (76 "missing" speaker_encoder.* keys is fine — the base encoder is
 used) → register speaker name→id in config.talker_config.spk_id → generate_custom_voice(...). No separate 
 speaker_embedding.safetensors needed. Params that helped: temperature=0.8, top_k=30, repetition_penalty=1.05, and cap 
 eos_token_id=[2150, 2157, 151670, 151673, 151645, 151643] for the ~0.5% of inferences that never EOS.

 On your dubbing/accent goal specifically

 One thing that bit us: the model learns both voice AND delivery style from the codec targets. If your targets are one style (e.g.
 read speech), you can't get a different style (e.g. whispered) just by swapping the speaker embedding at inference — the 
 embedding carries identity/prosody, but the acoustic style is baked into the targets. For cross-lingual accent transfer you'll 
 likely want the accent present in the training audio itself, not just in the reference. Worth setting that expectation.

 Credit: the instavar LoRA writeup documents bugs #1–2 and the LR/scale defaults; we just confirmed the sub-talker half of #2 and 
 added #3/#4. Good luck
